Formula/shadow-tls.rb

```diff
@@ -0,0 +1,37 @@
+class ShadowTls < Formula
+  desc "Proxy to expose real tls handshake to the firewall"
+  homepage "https://github.com/ihciah/shadow-tls"
+  url "https://github.com/ihciah/shadow-tls/archive/refs/tags/v0.2.8.tar.gz"
+  sha256 "f6c4094605a584e242a07776dae3879f1af9fe4ab020c81791349ceff9cbe0d8"
+  license "MIT"
+
+  depends_on "rustup-init" => :build
```

Comment:
Formulae should build with the stable rust toolchain

Reply:
Hi! Thanks for reviewing.
The component relies on [type_alias_impl_trait](https://rust-lang.github.io/rfcs/2515-type_alias_impl_trait.html) feature, which is unstable now.
I forked the rustfmt's formula(which is nightly-only too) from here: https://github.com/Homebrew/homebrew-core/blob/master/Formula/rustfmt.rb#L27
It is reproducable if using fixed version of nightly toolchain.

Reply:
> Then we should wait with this formula until that feature is stable.

Please note this feature started from more than 4 years ago and fully works from at least 1 year ago. The time it takes to wait for features to stabilize is unpredictable.
Another thing is, using nightly toolchain doesn't mean the compiled binary is unstable. It only means the api may change, the code may not compile on later toolchains. It is totally reproducable if using a fixed nightly version.
And rustfmt is in this repo too, I think it means rust nightly is acceptable and indeed works fine.

Reply:
> Please note this feature started from more than 4 years ago and fully works from at least 1 year ago. The time it takes to wait for features to stabilize is unpredictable.

That's not really an argument for Homebrew to start using binary dependencies

> Another thing is, using nightly toolchain doesn't mean the compiled binary is unstable. It only means the api may change, the code may not compile on later toolchains. It is totally reproducable if using a fixed nightly version.

But that pinned binary might stop working at any moment adding maintenance burden to the Homebrew project.

> And rustfmt is in this repo too, I think it means rust nightly is acceptable and indeed works fine.

It does not, it just means that an exception was made for rustfmt.

Reply:
> That's not really an argument for Homebrew to start using binary dependencies
> But that pinned binary might stop working at any moment adding maintenance burden to the Homebrew project.

Every building toolchain is binary dependency. Stable rust and nightly rust are downloaded from just the same place(rustup). If the nightly failed to download, the stable will fail too. And rustup promise all nightly toolchains are able to download([link](https://rust-lang.github.io/rustup/concepts/toolchains.html)). It won't expire, won't "suddenly stop working".

If multiple versions of gcc is acceptable, I think accepting fixed nightly rust is reasonable too. Given a fixed version number, the toolchain is frozened and always available.

> Using rustup-init is a violation of:
https://docs.brew.sh/Acceptable-Formulae#we-dont-like-install-scripts-that-download-unversioned-things

As I said, it is totally versioned. With a given version(for example nightly-2023-xx-xx) there will only be one corresponding toolchain, frozened, and always available. The purpose of this rule is simply to avoid something like `git clone xxxx` which does not reproducable and may compile fail any time. Here the case does not violate the exact rule and the idea behind it.

Reply:
> Every building toolchain is binary dependency. Stable rust and nightly rust are downloaded from just the same place(rustup).

It isn't. Stable rust is build from source.

> It won't expire, won't "suddenly stop working".

It'll stop working when Apple updates their OS in ways that are incompatible with that rust nightly.

> If multiple versions of gcc is acceptable, I think accepting fixed nightly rust is reasonable too. Given a fixed version number, the toolchain is frozened and always available.

If you compile that version of rust from source I totally agree.

> As I said, it is totally versioned.

So where is the checksum?

Reply:
> It'll stop working when Apple updates their OS in ways that are incompatible with that rust nightly.

I get what you mean now. You mean we don't want rustup-init(I still think it is versioned), you think it may fail on some macos(I think after some code reviewing we can determine whether it is compatible with all os we support including future). And you want something we control, so basicly the rust formula is a rewritten version of rustup-init but it is better.

So the core problem here is whether should we trust rustup-init. The trust means it is fully compatible and reproducible.

1. I didn't do code reviewing on rustup-init. But with my experience it works fine on both new and old macos versions. The rustup-init follows the same idea(versioned), if it is well-implemented, tested and brew can trust it, I think using rustup-init with fixed toolchain version will be acceptable.

2. If we not trust it, maybe the rust formula should provide some ways to use versioned nightly toolchain.

> So where is the checksum?

For example for nightly-2022-12-24 on aarch64-apple-darwin, the toolchain is at https://static.rust-lang.org/dist/2022-12-24/cargo-nightly-aarch64-apple-darwin.tar.gz and the checksum is https://static.rust-lang.org/dist/2022-12-24/cargo-nightly-aarch64-apple-darwin.tar.gz.sha256

Reply:
> For example for nightly-2022-12-24 on aarch64-apple-darwin, the toolchain is at https://static.rust-lang.org/dist/2022-12-24/cargo-nightly-aarch64-apple-darwin.tar.gz and the checksum is https://static.rust-lang.org/dist/2022-12-24/cargo-nightly-aarch64-apple-darwin.tar.gz.sha256

That's not used anywhere in the formula though, so it's still very much unchecksummed. They can reupload a compromised binary tomorrow and the Formula would download the new version without complaints.

> So the core problem here is whether should we trust rustup-init. The trust means it is fully compatible and reproducible.

No, the core problem is that Homebrew requires all software to be built from source and rustup isn't building the toolchain from source so it isn't meeting the requirements.

Reply:
So the solution maybe:
1. Create something like rust-nightly, installing rust from source.
2. Create workflow syncing new tar.gz version and its checksum.
3. For formulas depending on nightly, depend on rust-nightly.

This solution solves the "building from source" issue but it will be hard to use fixed nightly version as homebrew may not accept adding too many rust-nightly@2023-xx-xx into the git repo.

Another solution is to add feature to rustup-init formula(or create a new one), allowing and force another formula specify version and checksum to meet the requirement. But this solution breaks the "build from source" principle.

Except for not support and wait, do you have some ideas about making rust nightly available in homebrew?

> No, the core problem is that Homebrew requires all software to be built from source and rustup isn't building the toolchain from source so it isn't meeting the requirements.

It is still a trust problem though. We can break limitation for something that we can make sure safe(at least in rust programming).

Reply:
> Create some formula like rust-nightly, installing rust from source.

That would conflict with https://docs.brew.sh/Acceptable-Formulae#stable-versions. The real solution here is for rust projects not to depend on unstable features.

> Except for not support and wait, do you have some ideas about making rust nightly available in homebrew?

Not in homebrew/core where these requirements apply.

> It is still a trust problem though. We can break limitation for something that we can make sure safe(at least in rust programming).

Homebrew rules aren't based on rust though. They are based on problems maintainers have experienced over the years of building and maintaining packages.

Reply:
For rust(maybe other langs which can bootstrap), building from source doesn't mean purely from code. It will use pre-built binary for stage0. This behavior is difficult to clearly distinguish from directly copying the binary as the building process.

Reply:
Thanks for your time! I will close the issue.